File: crates/1bit-ingest/src/pack.rs
//! `pack` — TLV writer. Produces a `.1bl` from:
//!
//! * a trained `.gguf` (required, goes into `MODEL_GGUF` / 0x01),
//! * a `catalog.toml` (required, drives the CBOR header + ATTRIBUTION /
//!   LICENSE text sections),
//! * optional cover art (`COVER` / 0x02),
//! * optional lyrics bundle (`TRACK_LYRICS` / 0x03).
//!
//! See `docs/wiki/1bl-container-spec.md` §"Section order" for why the
//! GGUF is always written first and why the residual (if any) always
//! goes at the tail.

use std::fs::{File, read};
use std::io::{BufWriter, Write};
use std::path::Path;

use anyhow::{Context, Result, anyhow};
use byteorder::{LittleEndian, WriteBytesExt};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

use crate::{MAGIC, tag};

/// Strictly-typed mirror of the CBOR manifest the spec §"Header" requires.
/// Intentionally kept flat so a curator can read it straight out of their
/// `catalog.toml` without learning CBOR. We serialise the header via
/// `ciborium` on the way out and via `serde` on the way back in.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Manifest {
    pub v: String,
    pub catalog: String,
    pub title: String,
    pub artist: String,
    pub license: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub license_url: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub attribution: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source: Option<String>,
    pub created: String,
    pub tier: String,
    pub codec: CodecMeta,
    pub model: ModelMeta,
    pub tracks: Vec<Track>,
    #[serde(default)]
    pub residual_present: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub residual_sha256: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CodecMeta {
    pub audio: String,
    pub sample_rate: u32,
    pub channels: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelMeta {
    pub arch: String,
    pub params: u64,
    pub bpw: f64,
    /// Set by [`pack`] after hashing the embedded GGUF — the curator
    /// leaves this blank in `catalog.toml`.
    #[serde(default)]
    pub sha256: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Track {
    pub id: String,
    pub title: String,
    pub length_ms: u64,
    #[serde(default)]
    pub sha256: String,
}

/// Top-level `catalog.toml` schema. Lives in a separate struct from
/// [`Manifest`] so the curator's file doesn't have to include derived
/// fields (header version tag, model sha, residual_present flag).
#[derive(Debug, Clone, Deserialize)]
pub struct CatalogToml {
    pub catalog: String,
    pub title: String,
    pub artist: String,
    pub license: String,
    #[serde(default)]
    pub license_url: Option<String>,
    #[serde(default)]
    pub attribution: Option<String>,
    #[serde(default)]
    pub source: Option<String>,
    pub created: String,
    #[serde(default = "default_tier")]
    pub tier: String,
    pub codec: CodecMeta,
    pub model: ModelMeta,
    #[serde(default)]
    pub tracks: Vec<Track>,
    /// Inline license text. If absent, curator must provide
    /// `license_txt_path`.
    #[serde(default)]
    pub license_txt: Option<String>,
    #[serde(default)]
    pub license_txt_path: Option<String>,
    /// Inline attribution block. If absent and `attribution` is set,
    /// we synthesise a one-liner.
    #[serde(default)]
    pub attribution_txt: Option<String>,
}

fn default_tier() -> String {
    "lossy".into()
}

#[derive(Debug, Clone, Copy)]
pub struct PackSummary {
    pub section_count: usize,
    pub total_bytes: u64,
}

/// Pack a `.1bl`. The curator owns the `.toml` schema; this function owns
/// the byte layout.
pub fn pack(
    model_path: &Path,
    manifest_toml_path: &Path,
    cover_path: Option<&Path>,
    lyrics_path: Option<&Path>,
    out_path: &Path,
) -> Result<PackSummary> {
    // --- Load inputs ---------------------------------------------------
    let gguf_bytes = read(model_path)
        .with_context(|| format!("read model gguf: {}", model_path.display()))?;
    let toml_text = std::fs::read_to_string(manifest_toml_path)
        .with_context(|| format!("read catalog.toml: {}", manifest_toml_path.display()))?;
    let cat: CatalogToml = toml::from_str(&toml_text)
        .context("parse catalog.toml")?;

    // --- Derive fields -------------------------------------------------
    let gguf_sha = sha256_hex(&gguf_bytes);
    let license_txt = resolve_license_text(&cat, manifest_toml_path)?;
    let attribution_txt = cat.attribution_txt.clone().unwrap_or_else(|| {
        cat.attribution.clone().unwrap_or_else(|| cat.artist.clone())
    });

    let mut model = cat.model.clone();
    model.sha256 = gguf_sha;

    let manifest = Manifest {
        v: "0.1".into(),
        catalog: cat.catalog.clone(),
        title: cat.title.clone(),
        artist: cat.artist.clone(),
        license: cat.license.clone(),
        license_url: cat.license_url.clone(),
        attribution: cat.attribution.clone(),
        source: cat.source.clone(),
        created: cat.created.clone(),
        tier: cat.tier.clone(),
        codec: cat.codec.clone(),
        model,
        tracks: cat.tracks.clone(),
        residual_present: false,
        residual_sha256: None,
    };

    // --- Sections ------------------------------------------------------
    let cover_bytes = match cover_path {
        Some(p) => Some(read(p).with_context(|| format!("read cover: {}", p.display()))?),
        None => None,
    };
    let lyrics_bytes = match lyrics_path {
        Some(p) => Some(read(p).with_context(|| format!("read lyrics: {}", p.display()))?),
        None => None,
    };

    // --- Write out -----------------------------------------------------
    let out = File::create(out_path)
        .with_context(|| format!("create out: {}", out_path.display()))?;
    let mut w = HashingWriter::new(BufWriter::new(out));

    w.write_all(&MAGIC).context("write magic")?;

    let mut header_cbor = Vec::new();
    ciborium::into_writer(&manifest, &mut header_cbor)
        .map_err(|e| anyhow!("cbor encode header: {e}"))?;
    w.write_u32::<LittleEndian>(header_cbor.len() as u32)
        .context("write header length")?;
    w.write_all(&header_cbor).context("write header cbor")?;

    let mut section_count = 0usize;
    write_section(&mut w, tag::MODEL_GGUF, &gguf_bytes)?; section_count += 1;
    write_section(&mut w, tag::ATTRIBUTION_TXT, attribution_txt.as_bytes())?; section_count += 1;
    write_section(&mut w, tag::LICENSE_TXT, license_txt.as_bytes())?; section_count += 1;
    if let Some(b) = &cover_bytes { write_section(&mut w, tag::COVER, b)?; section_count += 1; }
    if let Some(b) = &lyrics_bytes { write_section(&mut w, tag::TRACK_LYRICS, b)?; section_count += 1; }

    // Footer = SHA-256 of everything written so far. Write the digest
    // directly (footer is NOT covered by itself).
    let digest = w.finalize_digest();
    let mut inner = w.into_inner();
    inner.write_all(&digest).context("write footer hash")?;
    inner.flush().context("flush output")?;

    let total_bytes = std::fs::metadata(out_path)
        .map(|m| m.len())
        .unwrap_or(0);
    Ok(PackSummary { section_count, total_bytes })
}

fn resolve_license_text(cat: &CatalogToml, toml_path: &Path) -> Result<String> {
    if let Some(t) = &cat.license_txt {
        return Ok(t.clone());
    }
    if let Some(rel) = &cat.license_txt_path {
        let base = toml_path.parent().unwrap_or(Path::new("."));
        let p = base.join(rel);
        let s = std::fs::read_to_string(&p)
            .with_context(|| format!("read license_txt_path: {}", p.display()))?;
        return Ok(s);
    }
    Err(anyhow!(
        "catalog.toml must set either `license_txt` (inline) or `license_txt_path` (sidecar). \
         Spec §\"Section tags\" lists LICENSE_TXT (0x05) as required."
    ))
}

pub(crate) fn write_section<W: Write>(
    w: &mut W,
    section_tag: u8,
    payload: &[u8],
) -> Result<()> {
    w.write_u8(section_tag).context("write section tag")?;
    w.write_u64::<LittleEndian>(payload.len() as u64).context("write section length")?;
    w.write_all(payload).context("write section payload")?;
    Ok(())
}

pub(crate) fn sha256_hex(bytes: &[u8]) -> String {
    let mut h = Sha256::new();
    h.update(bytes);
    let d = h.finalize();
    let mut s = String::with_capacity(64);
    const HEX: &[u8; 16] = b"0123456789abcdef";
    for &b in &d[..] {
        s.push(HEX[(b >> 4) as usize] as char);
        s.push(HEX[(b & 0xf) as usize] as char);
    }
    s
}

/// Writer that both passes bytes through to the inner `Write` and feeds
/// them into a running SHA-256, so the caller can append the digest as
/// a footer without re-reading the file.
pub(crate) struct HashingWriter<W: Write> {
    inner: W,
    hasher: Sha256,
}

impl<W: Write> HashingWriter<W> {
    pub(crate) fn new(inner: W) -> Self {
        Self { inner, hasher: Sha256::new() }
    }

    pub(crate) fn finalize_digest(&mut self) -> [u8; 32] {
        let taken = std::mem::replace(&mut self.hasher, Sha256::new());
        let out = taken.finalize();
        let mut arr = [0u8; 32];
        arr.copy_from_slice(&out);
        arr
    }

    pub(crate) fn into_inner(self) -> W { self.inner }
}

impl<W: Write> Write for HashingWriter<W> {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        let n = self.inner.write(buf)?;
        self.hasher.update(&buf[..n]);
        Ok(n)
    }
    fn flush(&mut self) -> std::io::Result<()> { self.inner.flush() }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample_toml(cat: &str) -> String {
        format!(r#"
catalog = "{cat}"
title = "Test Catalog"
artist = "Test Artist"
license = "CC0-1.0"
license_url = "https://creativecommons.org/publicdomain/zero/1.0/"
attribution = "Test Artist"
created = "2026-04-23T00:00:00Z"
tier = "lossy"
license_txt = "CC0 — public domain dedication."

[codec]
audio = "mimi-12hz"
sample_rate = 24000
channels = 2

[model]
arch = "bitnet-1p58"
params = 1048576
bpw = 1.58

[[tracks]]
id = "01"
title = "Track One"
length_ms = 60000
"#)
    }

    #[test]
    fn pack_writes_magic_header_and_sections() {
        let tmp = tempfile::tempdir().unwrap();
        let gguf = tmp.path().join("m.gguf");
        std::fs::write(&gguf, b"GGUF\0\0\0\0fake weights bytes here").unwrap();

        let tomlp = tmp.path().join("catalog.toml");
        std::fs::write(&tomlp, sample_toml("test-pack")).unwrap();

        let out = tmp.path().join("test.1bl");
        let summary = pack(&gguf, &tomlp, None, None, &out).unwrap();
        assert_eq!(summary.section_count, 3); // MODEL_GGUF + ATTRIBUTION + LICENSE
        assert!(summary.total_bytes > 32);

        let bytes = std::fs::read(&out).unwrap();
        assert_eq!(&bytes[..4], &MAGIC);
    }
}

File: cpp/ingest/src/pack.cpp
#include "onebit/ingest/ingest.hpp"

#include "cbor.hpp"
#include "onebit/ingest/sha256.hpp"

#include <toml++/toml.hpp>

#include <fstream>
#include <sstream>
#include <vector>

namespace onebit::ingest {

namespace {

namespace fs = std::filesystem;

[[nodiscard]] std::expected<std::vector<std::uint8_t>, IngestError>
read_all(const fs::path& p)
{
    std::ifstream in(p, std::ios::binary);
    if (!in) {
        return std::unexpected(IngestError::io(p, "open"));
    }
    in.seekg(0, std::ios::end);
    const auto sz = in.tellg();
    in.seekg(0, std::ios::beg);
    if (sz < 0) {
        return std::unexpected(IngestError::io(p, "tellg"));
    }
    std::vector<std::uint8_t> buf(static_cast<std::size_t>(sz));
    if (sz > 0) {
        in.read(reinterpret_cast<char*>(buf.data()),
                static_cast<std::streamsize>(sz));
        if (!in.good() && !in.eof()) {
            return std::unexpected(IngestError::io(p, "read"));
        }
    }
    return buf;
}

[[nodiscard]] std::expected<std::string, IngestError> read_text(const fs::path& p)
{
    auto bytes = read_all(p);
    if (!bytes) {
        return std::unexpected(bytes.error());
    }
    return std::string(reinterpret_cast<const char*>(bytes->data()), bytes->size());
}

void write_section(std::vector<std::uint8_t>& out,
                   std::uint8_t               tag,
                   std::span<const std::uint8_t> payload)
{
    out.push_back(tag);
    const auto len = static_cast<std::uint64_t>(payload.size());
    for (int i = 0; i < 8; ++i) {
        out.push_back(static_cast<std::uint8_t>(len >> (i * 8)));
    }
    out.insert(out.end(), payload.begin(), payload.end());
}

[[nodiscard]] detail::cbor::Value manifest_to_cbor(const Manifest& m)
{
    using onebit::ingest::detail::cbor::Array;
    using onebit::ingest::detail::cbor::Object;
    using onebit::ingest::detail::cbor::Value;

    Object obj;
    obj.emplace_back("v", Value{m.v});
    obj.emplace_back("catalog", Value{m.catalog});
    obj.emplace_back("title", Value{m.title});
    obj.emplace_back("artist", Value{m.artist});
    obj.emplace_back("license", Value{m.license});
    if (m.license_url) {
        obj.emplace_back("license_url", Value{*m.license_url});
    }
    if (m.attribution) {
        obj.emplace_back("attribution", Value{*m.attribution});
    }
    if (m.source) {
        obj.emplace_back("source", Value{*m.source});
    }
    obj.emplace_back("created", Value{m.created});
    obj.emplace_back("tier", Value{m.tier});

    Object codec;
    codec.emplace_back("audio", Value{m.codec.audio});
    codec.emplace_back("sample_rate",
                       Value{static_cast<std::int64_t>(m.codec.sample_rate)});
    codec.emplace_back("channels",
                       Value{static_cast<std::int64_t>(m.codec.channels)});
    obj.emplace_back("codec", Value{std::move(codec)});

    Object model;
    model.emplace_back("arch", Value{m.model.arch});
    model.emplace_back("params", Value{static_cast<std::int64_t>(m.model.params)});
    model.emplace_back("bpw", Value{m.model.bpw});
    model.emplace_back("sha256", Value{m.model.sha256});
    obj.emplace_back("model", Value{std::move(model)});

    Array tracks;
    for (const auto& t : m.tracks) {
        Object tt;
        tt.emplace_back("id", Value{t.id});
        tt.emplace_back("title", Value{t.title});
        tt.emplace_back("length_ms", Value{static_cast<std::int64_t>(t.length_ms)});
        tt.emplace_back("sha256", Value{t.sha256});
        tracks.emplace_back(Value{std::move(tt)});
    }
    obj.emplace_back("tracks", Value{std::move(tracks)});

    obj.emplace_back("residual_present", Value{m.residual_present});
    if (m.residual_sha256) {
        obj.emplace_back("residual_sha256", Value{*m.residual_sha256});
    }
    return Value{std::move(obj)};
}

[[nodiscard]] std::optional<std::string>
toml_str(const toml::table& t, std::string_view key)
{
    if (auto* v = t.get_as<std::string>(key)) {
        return v->get();
    }
    return std::nullopt;
}

} // namespace

// Public — also used by tests / future readers that need to interpret a
// hand-written toml file.
std::expected<CatalogToml, IngestError> parse_catalog_toml(std::string_view text)
{
    toml::table tbl;
    try {
        tbl = toml::parse(text);
    } catch (const toml::parse_error& e) {
        return std::unexpected(IngestError::toml(std::string{e.description()}));
    }
    CatalogToml c;
    auto require_str = [&](std::string_view key) -> std::expected<std::string, IngestError> {
        if (auto v = toml_str(tbl, key)) {
            return *v;
        }
        return std::unexpected(IngestError::toml(
            std::string{"missing required string key: "} + std::string{key}));
    };

    if (auto v = require_str("catalog"); v) c.catalog = std::move(*v); else return std::unexpected(v.error());
    if (auto v = require_str("title"); v) c.title = std::move(*v); else return std::unexpected(v.error());
    if (auto v = require_str("artist"); v) c.artist = std::move(*v); else return std::unexpected(v.error());
    if (auto v = require_str("license"); v) c.license = std::move(*v); else return std::unexpected(v.error());
    if (auto v = require_str("created"); v) c.created = std::move(*v); else return std::unexpected(v.error());

    c.license_url      = toml_str(tbl, "license_url");
    c.attribution      = toml_str(tbl, "attribution");
    c.source           = toml_str(tbl, "source");
    c.license_txt      = toml_str(tbl, "license_txt");
    c.license_txt_path = toml_str(tbl, "license_txt_path");
    c.attribution_txt  = toml_str(tbl, "attribution_txt");
    if (auto t = toml_str(tbl, "tier")) {
        c.tier = *t;
    }

    if (auto* codec = tbl.get_as<toml::table>("codec")) {
        if (auto v = toml_str(*codec, "audio")) {
            c.codec.audio = *v;
        }
        if (auto v = codec->get_as<std::int64_t>("sample_rate")) {
            c.codec.sample_rate = static_cast<std::uint32_t>(v->get());
        }
        if (auto v = codec->get_as<std::int64_t>("channels")) {
            c.codec.channels = static_cast<std::uint32_t>(v->get());
        }
    } else {
        return std::unexpected(IngestError::toml("missing [codec] table"));
    }

    if (auto* mdl = tbl.get_as<toml::table>("model")) {
        if (auto v = toml_str(*mdl, "arch")) {
            c.model.arch = *v;
        }
        if (auto v = mdl->get_as<std::int64_t>("params")) {
            c.model.params = static_cast<std::uint64_t>(v->get());
        }
        if (auto v = mdl->get_as<double>("bpw")) {
            c.model.bpw = v->get();
        } else if (auto v = mdl->get_as<std::int64_t>("bpw")) {
            c.model.bpw = static_cast<double>(v->get());
        }
    } else {
        return std::unexpected(IngestError::toml("missing [model] table"));
    }

    if (auto* tracks = tbl.get_as<toml::array>("tracks")) {
        for (auto&& el : *tracks) {
            if (auto* tt = el.as_table()) {
                Track t;
                if (auto v = toml_str(*tt, "id")) {
                    t.id = *v;
                }
                if (auto v = toml_str(*tt, "title")) {
                    t.title = *v;
                }
                if (auto v = tt->get_as<std::int64_t>("length_ms")) {
                    t.length_ms = static_cast<std::uint64_t>(v->get());
                }
                if (auto v = toml_str(*tt, "sha256")) {
                    t.sha256 = *v;
                }
                c.tracks.push_back(std::move(t));
            }
        }
    }

    return c;
}

std::expected<PackSummary, IngestError>
pack(const fs::path&                                     model_path,
     const fs::path&                                     manifest_toml_path,
     const std::optional<fs::path>&                      cover_path,
     const std::optional<fs::path>&                      lyrics_path,
     const fs::path&                                     out_path)
{
    auto gguf_bytes = read_all(model_path);
    if (!gguf_bytes) {
        return std::unexpected(gguf_bytes.error());
    }
    auto toml_text = read_text(manifest_toml_path);
    if (!toml_text) {
        return std::unexpected(toml_text.error());
    }
    auto cat = parse_catalog_toml(*toml_text);
    if (!cat) {
        return std::unexpected(cat.error());
    }

    // Resolve license text — inline first, sidecar second.
    std::string license_txt;
    if (cat->license_txt) {
        license_txt = *cat->license_txt;
    } else if (cat->license_txt_path) {
        const auto base = manifest_toml_path.parent_path();
        const auto p    = base / *cat->license_txt_path;
        auto       t    = read_text(p);
        if (!t) {
            return std::unexpected(t.error());
        }
        license_txt = *t;
    } else {
        return std::unexpected(IngestError::invalid(
            "catalog.toml must set license_txt (inline) or license_txt_path (sidecar)"));
    }

    const std::string attribution_txt =
        cat->attribution_txt ? *cat->attribution_txt
                             : (cat->attribution ? *cat->attribution : cat->artist);

    Manifest m;
    m.v                = "0.1";
    m.catalog          = cat->catalog;
    m.title            = cat->title;
    m.artist           = cat->artist;
    m.license          = cat->license;
    m.license_url      = cat->license_url;
    m.attribution      = cat->attribution;
    m.source           = cat->source;
    m.created          = cat->created;
    m.tier             = cat->tier;
    m.codec            = cat->codec;
    m.model            = cat->model;
    m.model.sha256     = detail::sha256_hex(*gguf_bytes);
    m.tracks           = cat->tracks;
    m.residual_present = false;

    std::optional<std::vector<std::uint8_t>> cover_bytes;
    if (cover_path) {
        auto b = read_all(*cover_path);
        if (!b) {
            return std::unexpected(b.error());
        }
        cover_bytes = std::move(*b);
    }
    std::optional<std::vector<std::uint8_t>> lyrics_bytes;
    if (lyrics_path) {
        auto b = read_all(*lyrics_path);
        if (!b) {
            return std::unexpected(b.error());
        }
        lyrics_bytes = std::move(*b);
    }

    // Build the file in memory; the sizes here are bounded by the
    // GGUF (single-digit GB upper bound for our use). Streaming would
    // be a follow-up if curators ship multi-GB models.
    std::vector<std::uint8_t> out;
    out.reserve(gguf_bytes->size() + 4096);
    out.insert(out.end(), MAGIC.begin(), MAGIC.end());

    auto cbor_bytes = detail::cbor::encode(manifest_to_cbor(m));
    if (cbor_bytes.size() > 0xFFFFFFFFULL) {
        return std::unexpected(IngestError::header_too_large(cbor_bytes.size()));
    }
    const auto h_len = static_cast<std::uint32_t>(cbor_bytes.size());
    for (int i = 0; i < 4; ++i) {
        out.push_back(static_cast<std::uint8_t>(h_len >> (i * 8)));
    }
    out.insert(out.end(), cbor_bytes.begin(), cbor_bytes.end());

    std::size_t section_count = 0;
    write_section(out, tag::MODEL_GGUF, *gguf_bytes);
    ++section_count;
    write_section(out, tag::ATTRIBUTION_TXT,
                  std::span<const std::uint8_t>{
                      reinterpret_cast<const std::uint8_t*>(attribution_txt.data()),
                      attribution_txt.size()});
    ++section_count;
    write_section(out, tag::LICENSE_TXT,
                  std::span<const std::uint8_t>{
                      reinterpret_cast<const std::uint8_t*>(license_txt.data()),
                      license_txt.size()});
    ++section_count;
    if (cover_bytes) {
        write_section(out, tag::COVER, *cover_bytes);
        ++section_count;
    }
    if (lyrics_bytes) {
        write_section(out, tag::TRACK_LYRICS, *lyrics_bytes);
        ++section_count;
    }

    // Footer — SHA-256 over everything written so far.
    const auto digest = detail::sha256(out);
    out.insert(out.end(), digest.begin(), digest.end());

    std::ofstream of(out_path, std::ios::binary | std::ios::trunc);
    if (!of) {
        return std::unexpected(IngestError::io(out_path, "create"));
    }
    of.write(reinterpret_cast<const char*>(out.data()),
             static_cast<std::streamsize>(out.size()));
    if (!of.good()) {
        return std::unexpected(IngestError::io(out_path, "write"));
    }
    of.flush();

    return PackSummary{section_count, static_cast<std::uint64_t>(out.size())};
}

std::string IngestError::what() const
{
    return std::visit(
        [](const auto& e) -> std::string {
            using T = std::decay_t<decltype(e)>;
            if constexpr (std::is_same_v<T, ErrIo>) {
                return "io: " + e.path.string() + ": " + e.message;
            } else if constexpr (std::is_same_v<T, ErrBadMagic>) {
                std::ostringstream o;
                o << "bad magic bytes (got";
                for (auto b : e.got) {
                    o << ' ' << std::hex << static_cast<int>(b);
                }
                o << ')';
                return o.str();
            } else if constexpr (std::is_same_v<T, ErrBadVersion>) {
                std::ostringstream o;
                o << "unsupported version byte: " << std::hex
                  << static_cast<int>(e.version);
                return o.str();
            } else if constexpr (std::is_same_v<T, ErrFooterHashMismatch>) {
                return "footer hash mismatch (file corrupt or truncated)";
            } else if constexpr (std::is_same_v<T, ErrTruncated>) {
                return std::string{"truncated file: "} + e.what;
            } else if constexpr (std::is_same_v<T, ErrHeaderTooLarge>) {
                return "header too large: " + std::to_string(e.bytes) +
                       " bytes (cap is 16 MiB)";
            } else if constexpr (std::is_same_v<T, ErrCbor>) {
                return "cbor decode: " + e.message;
            } else if constexpr (std::is_same_v<T, ErrToml>) {
                return "toml: " + e.message;
            } else if constexpr (std::is_same_v<T, ErrInvalid>) {
                return e.message;
            }
        },
        v_);
}

} // namespace onebit::ingest
